### PR TITLE
Patch 3: fix #3, fix #8

### LINKED ISSUE #3: Sync *.js with upstream n2o

### LINKED ISSUE #8: Cyrillic encoding error
server/client OS: Windows 7 x64.

How to reproduce: 
1. run n2o.exe
2. run server with sample/client.html
3. connect as "user".
4. type "ыыы" (without quotes) and press Enter.

ER: "ыыы" on client side.
AR: 
- "KKK" on client-side
- [AtomTerm "MSG",AtomTerm "KKK"] as a server log.

### PR DESCRIPTION
**Pull request by swamp-agr**

### CODE REVIEW

File: sample/client.js [hunk 1]
```diff
@@ -30,14 +30,14 @@ function onMessage(event) {
                 eval(x)
             }
             else if (typeof x == 'object'
-                && x.type == 'Tuple' 
-                && x.value[0].length == 3
-                && x.value[0][0] == 'io') 
+                && x.t == 104
```
Comment (nponeccop) [inline sample/client.js:33]: Should be == TUPLE (w/o quotes)

File: src/n2o.hs [hunk 2]
```diff
@@ -53,4 +59,5 @@ handle state entry [AtomTerm "N2O_DISCONNECT"]
 
 handle _state _entry _ = putStrLn "Protocol violation"
 
-
+lazyDecodeUtf8 = decodeUtf8 . toStrict
+lazyEncodeUtf8 = fromStrict . encodeUtf8
```
Comment (nponeccop) [inline src/n2o.hs:63]: Can't you use the `fromString` trick instead?

**Comment by nponeccop** [comment]
Issue #8 is only partially fixed - cyrillic login encoding problem remains

File: src/n2o.hs [hunk 3]
```diff
@@ -17,30 +20,33 @@ main = runServer "0.0.0.0" 9160 handle loggedOff
...
         if ce 
             then alert entry "User already exists"
             else do
                 send entry $ call "joinSession" ""
-                setState state (eSocketId entry) $ Just $ fromString name
+                setState state (eSocketId entry) $ Just dname
                 clients <- loggedOn state
                 let foo = concatMap ((\x -> "<li>" <> x <> "</li>") . fromJust . eUser) clients
-                broadcast (call "$('#users').html" (fromString foo) <> call "log" (fromString name <> " joined")) clients
+                broadcast (call "$('#users').html" (fromString foo) <> call "log" ((fromString dname) <> " joined")) clients
```
Comment (nponeccop) [inline src/n2o.hs:47]: Reduntant parentheses

File: sample/client.js (hunk 1 above)
Comment (swamp-agr) [inline sample/client.js:33]:
```
Uncaught ReferenceError: TUPLE is not defined
    at <anonymous>:2:1
    at Object.InjectedScript._evaluateOn (<anonymous>:875:140)
    at Object.InjectedScript._evaluateAndWrap (<anonymous>:808:34)
    at Object.InjectedScript.evaluate (<anonymous>:664:21)
```

File: src/n2o.hs (hunk 3 above)
Comment (swamp-agr) [inline src/n2o.hs:47]: Thank you. With username it will be fixed.

**Pull request merged**